File: android/src/main/kotlin/com/voximplant/voximplant_kit_chat/VoximplantKitChatPlugin.kt
/*
 * Copyright (c) 2011 - 2026, Voximplant, Inc. All rights reserved.
 */

package com.voximplant.voximplant_kit_chat

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.annotation.UiThread
import com.voximplant.android.kit.chat.core.exceptions.KitConnectionRequiredException
import com.voximplant.android.kit.chat.core.exceptions.KitInternalException
import com.voximplant.android.kit.chat.core.exceptions.KitTimeoutException
import com.voximplant.android.kit.chat.ui.KitChatUi
import com.voximplant.android.kit.chat.ui.model.AuthorizationError
import com.voximplant.android.kit.chat.ui.model.ClientData
import com.voximplant.android.kit.chat.ui.model.Region
import io.flutter.embedding.engine.plugins.FlutterPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/** VoximplantKitChatPlugin */
class VoximplantKitChatPlugin : FlutterPlugin, VoximplantKitChatApi {
    private var applicationContext: Context? = null
    private var kitChatUi: KitChatUi? = null
    private var flutterApi: VoximplantKitChatFlutterApi? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    private val coroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = flutterPluginBinding.applicationContext
        flutterApi = VoximplantKitChatFlutterApi(flutterPluginBinding.binaryMessenger)
        VoximplantKitChatApi.setUp(flutterPluginBinding.binaryMessenger, this)
    }

    override fun initialize(credentials: KitChatCredentials) {
        kitChatUi = KitChatUi(
            context = applicationContext ?: throw FlutterError(
                code = "INITIALIZATION_FAILED",
                message = "Android context is not attached",
            ),
            accountRegion = credentials.region.toRegion(),
            channelUuid = credentials.channelUuid,
            token = credentials.token,
            clientId = credentials.clientId
        ).apply {
            onAuthorizationError = { error ->
                Log.e(TAG, "KitChat authorization error: $error")
                mainHandler.post {
                    flutterApi?.onAuthorizationError(error.toKitChatAuthorizationError()) { result ->
                        result.exceptionOrNull()?.let { callbackError ->
                            Log.e(TAG, "Failed to send auth error to Flutter: $callbackError")
                        }
                    }
                }
            }
        }
    }

    override fun applyCustomization(customization: KitChatCustomization) {
        customization.colorScheme?.let { colorScheme ->
            KitChatUi.colorScheme = colorScheme.toKitChatColorScheme()
        }
    }

    @UiThread
    override fun openChat() {
        kitChatUi?.startActivity() ?: throw FlutterError(
            code = "NOT_INITIALIZED",
            message = "KitChatUi is not initialized",
        )
    }

    /** Android does not track whether the chat is currently open. */
    override fun isChatVisible(): Boolean = false

    override fun setClientData(data: KitChatClientData, callback: (Result<Unit>) -> Unit) {
        val currentKitChatUi = kitChatUi
        if (currentKitChatUi == null) {
            callback(Result.failure(FlutterError("NOT_INITIALIZED", "KitChatUi is not initialized")))
            return
        }

        val clientData = ClientData(
            displayName = data.displayName,
            avatarUrl = data.avatarUrl,
            email = data.email,
            phone = data.phone,
            language = data.language
        )
        coroutineScope.launch {
            currentKitChatUi.setClientData(clientData).onSuccess {
                callback(Result.success(Unit))
            }.onFailure { exception ->
                val code = when (exception) {
                    is IllegalArgumentException -> "ILLEGAL_ARGUMENT"
                    is KitConnectionRequiredException -> "CONNECTION_REQUIRED"
                    is KitTimeoutException -> "TIMEOUT"
                    is KitInternalException -> "INTERNAL"
                    else -> "UNKNOWN"
                }
                callback(Result.failure(FlutterError(code, exception.message)))
            }
        }
    }

    override fun registerPushToken(token: String, callback: (Result<Unit>) -> Unit) {
        val currentKitChatUi = kitChatUi
        if (currentKitChatUi == null) {
            callback(Result.failure(FlutterError("NOT_INITIALIZED", "KitChatUi is not initialized")))
            return
        }

        coroutineScope.launch {
            currentKitChatUi.registerPushToken(token).onSuccess {
                callback(Result.success(Unit))
            }.onFailure { exception ->
                val code = when (exception) {
                    is IllegalArgumentException -> "ILLEGAL_ARGUMENT"
                    is KitConnectionRequiredException -> "CONNECTION_REQUIRED"
                    is KitTimeoutException -> "TIMEOUT"
                    is KitInternalException -> "INTERNAL"
                    else -> "UNKNOWN"
                }
                callback(Result.failure(FlutterError(code, exception.message)))
            }
        }
    }

    override fun unregisterPushToken(token: String, callback: (Result<Unit>) -> Unit) {
        val currentKitChatUi = kitChatUi
        if (currentKitChatUi == null) {
            callback(Result.failure(FlutterError("NOT_INITIALIZED", "KitChatUi is not initialized")))
            return
        }

        coroutineScope.launch {
            currentKitChatUi.unregisterPushToken(token).onSuccess {
                callback(Result.success(Unit))
            }.onFailure { exception ->
                val code = when (exception) {
                    is IllegalArgumentException -> "ILLEGAL_ARGUMENT"
                    is KitConnectionRequiredException -> "CONNECTION_REQUIRED"
                    is KitTimeoutException -> "TIMEOUT"
                    is KitInternalException -> "INTERNAL"
                    else -> "UNKNOWN"
                }
                callback(Result.failure(FlutterError(code, exception.message)))
            }
        }
    }

    override fun handlePush(payload: Map<String, String>, callback: (Result<Unit>) -> Unit) {
        val currentKitChatUi = kitChatUi
        if (currentKitChatUi == null) {
            callback(Result.failure(FlutterError("NOT_INITIALIZED", "KitChatUi is not initialized")))
            return
        }

        try {
            currentKitChatUi.handlePush(payload)
            callback(Result.success(Unit))
        } catch (exception: Exception) {
            callback(Result.failure(FlutterError("HANDLE_PUSH_FAILED", exception.message)))
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        VoximplantKitChatApi.setUp(binding.binaryMessenger, null)
        coroutineScope.cancel()
        applicationContext = null
        kitChatUi = null
        flutterApi = null
    }

    private fun KitChatRegion.toRegion(): Region = when (this) {
        KitChatRegion.RU -> Region.RU
        KitChatRegion.RU2 -> Region.RU_2
        KitChatRegion.EU -> Region.EU
        KitChatRegion.US -> Region.US
        KitChatRegion.BR -> Region.BR
        KitChatRegion.KZ -> Region.KZ
    }

    private fun AuthorizationError.toKitChatAuthorizationError(): KitChatAuthorizationError = when (this) {
        is AuthorizationError.InvalidChannelUuid -> KitChatAuthorizationError.INVALID_CHANNEL_UUID
        is AuthorizationError.InvalidToken -> KitChatAuthorizationError.INVALID_TOKEN
        is AuthorizationError.InvalidClientId -> KitChatAuthorizationError.INVALID_CLIENT_ID
        else -> KitChatAuthorizationError.UNKNOWN
    }

    private companion object {
        const val TAG = "VoximplantKitChat"
    }
}

File: ios/voximplant_kit_chat/Sources/voximplant_kit_chat/VoximplantKitChatPlugin.swift
import Flutter
import UIKit
import VoximplantKitChatUI

public final class VoximplantKitChatPlugin: NSObject, FlutterPlugin, VoximplantKitChatApi, VIKitChatViewControllerDelegate {
    private var viKitChatUI: VIKitChatUI?
    private var flutterApi: VoximplantKitChatFlutterApi?
    private var kitChatCustomization = VIKitChatUICustomization()


    private var isDevelopmentPushToken: Bool {
        #if DEBUG
        true
        #else
        false
        #endif
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = VoximplantKitChatPlugin()
        instance.flutterApi = VoximplantKitChatFlutterApi(binaryMessenger: registrar.messenger())
        VoximplantKitChatApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
    }

    func initialize(credentials: KitChatCredentials) throws {
        guard
            let viKitChatUI = VIKitChatUI(
                accountRegion: credentials.region.viRegion,
                channelUuid: credentials.channelUuid,
                token: credentials.token,
                clientId: credentials.clientId
            )
        else {
            throw PigeonError(
                code: "INITIALIZATION_FAILED",
                message: "Failed to initialize VIKitChatUI with provided credentials",
                details: nil
            )
        }

        self.viKitChatUI = viKitChatUI
    }

    func applyCustomization(customization: KitChatCustomization) throws {
        var kitCustomization = VIKitChatUICustomization()
        if let colorScheme = customization.colorScheme {
            kitCustomization.colorScheme = colorScheme.viColorScheme
        }
        if let customizableStrings = customization.customizableStringsIos {
            kitCustomization.strings = customizableStrings.viCustomizableStrings
        }
        if let customizableIcons = customization.customizableIconsIos {
            kitCustomization.icons = customizableIcons.viCustomizableIcons
        }
        self.kitChatCustomization = kitCustomization
    }

    func openChat() throws {
        guard let viKitChatUI = self.viKitChatUI else {
            throw PigeonError(code: "NOT_INITIALIZED", message: "KitChatUI is not initialized", details: nil)
        }

        let rootViewController = try getRootViewController()
        let navigationController = try getNavigationController(from: rootViewController)

        guard let kitChatViewController = VIKitChatViewController(
            id: viKitChatUI.id,
            customization: kitChatCustomization
        )
        else {
            throw PigeonError(
                code: "MISSING_KIT_CHAT_VIEW_CONTROLLER",
                message: "Unable to create chat view controller",
                details: nil
            )
        }

        kitChatViewController.delegate = self
        navigationController.pushViewController(kitChatViewController, animated: true)
    }

    func isChatVisible() throws -> Bool {
        Self.isChatVisible()
    }

    func setClientData(data: KitChatClientData, completion: @escaping (Result<Void, Error>) -> Void) {
        guard let viKitChatUI = self.viKitChatUI else {
            completion(.failure(PigeonError(code: "NOT_INITIALIZED", message: "KitChatUI is not initialized", details: nil)))
            return
        }

        let clientData = VIClientData(
            displayName: data.displayName,
            phone: data.phone,
            avatarUrl: data.avatarUrl,
            email: data.email,
            language: data.language
        )
        viKitChatUI.setClientData(clientData) { error in
            if let error {
                let code: String
                switch error {
                case .emptyData, .invalidEmail:
                    code = "ILLEGAL_ARGUMENT"
                case .networkIssues:
                    code = "CONNECTION_REQUIRED"
                case .timeout:
                    code = "TIMEOUT"
                case .internalError:
                    code = "INTERNAL"
                @unknown default:
                    code = "UNKNOWN"
                }
                completion(.failure(PigeonError(
                    code: code,
                    message: "Failed to set client data: \(error)",
                    details: nil
                )))
            } else {
                completion(.success(()))
            }
        }
    }

    func registerPushToken(token: String, completion: @escaping (Result<Void, Error>) -> Void) {
        guard let viKitChatUI = self.viKitChatUI else {
            completion(.failure(PigeonError(code: "NOT_INITIALIZED", message: "KitChatUI is not initialized", details: nil)))
            return
        }

        guard let pushTokenData = Data(hex: token) else {
            completion(.failure(PigeonError(
                code: "ILLEGAL_ARGUMENT",
                message: "Invalid push token",
                details: nil
            )))
            return
        }
        viKitChatUI.registerPushToken(pushTokenData, isDevelopment: isDevelopmentPushToken) { error in
            if let error {
                let code: String
                switch error {
                case .emptyPushToken:
                    code = "ILLEGAL_ARGUMENT"
                case .networkIssues:
                    code = "CONNECTION_REQUIRED"
                case .timeout:
                    code = "TIMEOUT"
                case .internalError:
                    code = "INTERNAL"
                @unknown default:
                    code = "UNKNOWN"
                }
                completion(.failure(PigeonError(
                    code: code,
                    message: "Failed to register push token: \(error)",
                    details: nil
                )))
            } else {
                completion(.success(()))
            }
        }
    }

    func unregisterPushToken(token: String, completion: @escaping (Result<Void, Error>) -> Void) {
        guard let viKitChatUI = self.viKitChatUI else {
            completion(.failure(PigeonError(code: "NOT_INITIALIZED", message: "KitChatUI is not initialized", details: nil)))
            return
        }
        guard let pushTokenData = Data(hex: token) else {
            completion(.failure(PigeonError(
                code: "ILLEGAL_ARGUMENT",
                message: "Invalid push token",
                details: nil
            )))
            return
        }
        viKitChatUI.unregisterPushToken(pushTokenData, isDevelopment: isDevelopmentPushToken) { error in
            if let error {
                let code: String
                switch error {
                case .emptyPushToken:
                    code = "ILLEGAL_ARGUMENT"
                case .networkIssues:
                    code = "CONNECTION_REQUIRED"
                case .timeout:
                    code = "TIMEOUT"
                case .internalError:
                    code = "INTERNAL"
                @unknown default:
                    code = "UNKNOWN"
                }
                completion(.failure(PigeonError(
                    code: code,
                    message: "Failed to unregister push token: \(error)",
                    details: nil
                )))
            } else {
                completion(.success(()))
            }
        }
    }

    func handlePush(payload: [String: String], completion: @escaping (Result<Void, Error>) -> Void) {
        // Kept for API compatibility with Android implementation
        completion(.success(()))
    }

    public func kitChatViewController(
        _ kitChatViewController: VIKitChatViewController,
        didFailToAuthorizeWithError error: VIAuthorizationError
    ) {
        print("VoximplantKitChat authorization error: \(error)")

        let authError = error.kitChatAuthorizationError
        DispatchQueue.main.async {
            guard let flutterApi = self.flutterApi else {
                print("VoximplantKitChat Flutter API is not set up; authorization error was not forwarded")
                return
            }

            flutterApi.onAuthorizationError(error: authError) { result in
                switch result {
                case .success:
                    print("VoximplantKitChat forwarded authorization error to Flutter: \(authError)")
                case .failure(let error):
                    print("VoximplantKitChat failed to send auth error to Flutter: \(error)")
                }
            }
        }
    }

    private func getRootViewController() throws -> UIViewController {
        guard let rootViewController = UIApplication.shared.firstWindow?.rootViewController else {
            throw PigeonError(
                code: "MISSING_VIEW_CONTROLLER",
                message: "Unable to find root view controller",
                details: nil
            )
        }
        return rootViewController
    }

    private func getNavigationController(from rootViewController: UIViewController) throws -> UINavigationController {
        guard
            let navigationController = (rootViewController as? UINavigationController)
                ?? rootViewController.navigationController
        else {
            throw PigeonError(
                code: "MISSING_VIEW_CONTROLLER",
                message: "Navigation controller is required to open chat",
                details: nil
            )
        }
        return navigationController
    }
}

private extension KitChatRegion {
    var viRegion: VIRegion {
        switch self {
        case .ru:
            return .ru
        case .ru2:
            return .ru2
        case .eu:
            return .eu
        case .us:
            return .us
        case .br:
            return .br
        case .kz:
            return .kz
        }
    }
}

private extension VIAuthorizationError {
    var kitChatAuthorizationError: KitChatAuthorizationError {
        switch self {
        case .invalidChannelUuid: .invalidChannelUuid
        case .invalidToken: .invalidToken
        case .invalidClientId: .invalidClientId
        @unknown default: .unknown
        }
    }
}

private extension Data {
    init?(hex: String) {
        guard hex.count % 2 == 0 else {
            return nil
        }
        let bytes = stride(from: 0, to: hex.count, by: 2).compactMap {
            UInt8(hex[hex.index(hex.startIndex, offsetBy: $0)..<hex.index(hex.startIndex, offsetBy: $0 + 2)], radix: 16)
        }
        guard bytes.count == hex.count / 2 else {
            return nil
        }
        self.init(bytes)
    }
}

public extension VoximplantKitChatPlugin {
    /// Returns `true` if the KitChat screen is currently visible to the user.
    static func isChatVisible() -> Bool {
        UIApplication.shared.firstWindow?.visibleViewController is VIKitChatViewController
    }
}

private extension UIApplication {
    var firstWindow: UIWindow? {
        connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .first(where: { $0.activationState == .foregroundActive })?
            .windows
            .first(where: \.isKeyWindow)
    }
}

private extension UIWindow {
    var visibleViewController: UIViewController? {
        Self.findVisibleViewController(rootViewController)
    }

    private static func findVisibleViewController(_ controller: UIViewController?) -> UIViewController? {
        guard let controller else { return nil }
        if let presented = controller.presentedViewController {
            return findVisibleViewController(presented)
        }
        if let nav = controller as? UINavigationController {
            return findVisibleViewController(nav.visibleViewController)
        }
        if let tab = controller as? UITabBarController {
            return findVisibleViewController(tab.selectedViewController)
        }
        return controller
    }
}
